### ISSUE TITLE
Invalid MEX-file No such file or directory

### ISSUE THREAD

**Issue by Qianli-ion** [kyamagu/mexopencv#260]
Greetings!

Thanks for this awesome package! I tried to run the facedect.m and my matlab returns this error. 

> Invalid MEX-file '/eecf/cbcsl/data100/opencv-qianli/mexopencv-2.4/+cv/private/CascadeClassifier_.mexa64':
> libopencv_calib3d.so.2.4: cannot open shared object file: No such file or directory
> 
> Error in cv.CascadeClassifier (line 33)
>             this.id = CascadeClassifier_(filename);
> 
> Error in facedetect (line 10)
> classifier = cv.CascadeClassifier(xml_file);

My OpenCV version is 2.4.13 and mexopencv version is 2.4. The installation procedure of mine case is a little bit different. The access permission on my machine is quite limited. /usr/local/ is not writeable so in the OpenCV cmake configuration I set the CMAKE_INSTALL_PREFIX to a local folder(opencv-2.4.13/build2) which I have access to. The --cflags and --libs of opencv returned by pkg-config are also not correct and the opencv.pc in the system's pkgconfig folder cannot be modified as well. So I set the C_FLAGS and CV_LDFLAGS in the mexopencv Makefile to the Cflags and Libs in the build2/libs/pkg-config/opencv.pc, respectively. The g++ version on this machine does not support -std=c++11 so I changed all the c++11 to c++0x in the mex_C++_glnxa64.xml. For more details about Makefile, please refer to the attached Makefile_modified.txt.
[Makefile_modified.txt](https://github.com/kyamagu/mexopencv/files/345296/Makefile_modified.txt)

The compilation seems successful. But when I run the above mentioned code, it returns the above error. I tried run the same code after entering setenv('LD_LIBRARY_PATH',['/eecf/cbcsl/data100/opencv-qianli/opencv-2.4.13/build2/lib' pathsep() getenv('LD_LIBRARY_PATH')]); in matlab command window. But the error is the same. 

Any ideas? Thank you very much!

**Comment by amroamroamro** [comment]
It is possible to compile mexopencv against opencv installed in a custom location without modifying the makefile:

- tell `pkg-config` to look for the `opencv.pc` file in a different location by setting an environment variable: `export PKG_CONFIG_PATH=~/cv/build/lib/pkgconfig:$PKG_CONFIG_PATH`. You can then check that the output of `pkg-config --cflags --libs opencv` is indeed using your modified file.

- pass any variable you want to modify directly in the make call, e.g: `make MATLABDOR=/opt/local/matlab MEXEXT=mexa64`

- finally use the `LD_LIBRARY_PATH` environment variable to configure where OpenCV shared library are loaded from. So first set `export LD_LIBRARY_PATH=~/cv/build/lib:$LD_LIBRARY_PATH` then launch matlab from the same terminal.

Also if I'm not mistaken, we haven't used any C++11 features in mexopencv, so you should be able to build it with a  regular c++03 compiler.

**Comment by Qianli-ion** [comment]
Thanks a lot for your reply!

I was working on the first suggestion you gave. After entering the `export PKG_CONFIG_PATH=~/cv/build/lib/pkgconfig:$PKG_CONFIG_PATH` and then `pkg-config --libs opencv` I still got

>  -lcxcore -lcv -lhighgui -lcvaux -lml 

which is the libs in the original opencv.pc in the system's pkgconfig folder. 

Some details, the first several times I tried `export` command the terminal returned 

> PKG_CONFIG_PATH: Undefined variable.

I did a little bit research. It seems like I was running tcsh shell. So I used `bash -c 'export...' `. And after the `export` command, I tried to `bash -c 'echo $PKG_CONFIG_PATH'` but the terminal gave me a blank line.

**Comment by Qianli-ion** [comment]
Besides, inspired from your last suggestion, I didn't do `export LD_LIBRARY_PATH=...` since previously when I tried the command the terminal says export: command not found. I suspect it is the same reason as in my last comment. I will try to do this one in a bash shell and run the matlab from the same terminal. 

I'll keep you posted. Thanks again!

**Comment by kyamagu** [comment]
@Qianli-ion FYI, [here is the shell differences](https://web.fe.up.pt/~jmcruz/etc/unix/sh-vs-csh.html)

**Comment by Qianli-ion** [comment]
@kyamagu 山口さん (hopefully I spelled it correctly) thanks a lot for your information! This is very helpful. After doing the `export LD_LIBRARY_PATH` in bash it works like a charm. I'm still trying to learn the shell difference and the reason that the pkg-config didn't gave me the local version of my OpenCV. But I think this question is out of the mexopencv so I think work on my own would be a better idea. Thanks so much for the help from both of you @amroamroamro @kyamagu.

**Comment by amroamroamro** [comment]
@Qianli-ion 
To set an environment variable in the shell you have to `export` it. The "scope" of exported shell variables is the process and its children. That means when you run `bash -c 'export VAR=value'` the changes are lost immediately after the command finishes (a new child process is started to run the command, and gets destroyed once it finishes).

For example compare these two:

``` sh
$ export AAA=1
$ echo $AAA    # 1
```
``` sh
$ bash -c 'export BBB=2'
$ echo $BBB    # empty
```

In this case, simply start an interactive bash shell as usual (just type `bash` in your current terminal), and execute all your commands in that shell (export and run make), then launch matlab from the same shell, that way it inherits any environment variables you set (of importance is the LD_LIBRARY_PATH).

Of course the `export` syntax is specific to sh/bash. You can instead work in the C-shell if you prefer, but you have to use equivalent for that shell as kyamagu showed.

**Comment by Qianli-ion** [comment]
Oh...So that is the reason why the terminal returned a blank line after my `echo $PKG_CONFIG_PATH` command. When I used `bash -c 'export PKG_CONFIG_PATH=...'`the terminal create a little bash shell and assign the directory I put in `...` to `PKG_CONFIG_PATH` but once that line is finished the little bash shell got destroyed and I lose the `PKG_CONFIG_PATH` I just set. 

Yes I tried getting into the interactive bash shell as you described and the `pkg-config --libs opencv` returns exact the files in my local opencv. That's perfect! I don't have to modify Makefile actually. (That's a little bit sad frankly speaking, since figuring out how to modify Makefile to get around the permission limits took me quite a bit time)

By the way, the C++11 I mentioned in the first comment is not from the mexopencv package, but from the default matlab setting I think. When I run `make` without changing mex_C++_glnxa64.xml, it gave me a 

> cc1plus: error: unrecognized command line option -std=c++11

Then I added a `-v` flag in the `$(MEX) -c $(C_FLAGS) $< -outdir $(LIBDIR)` in the Makefile and found out the directory of my mex_C++_glnxa64.xml. After that I changed the c++11 in that file to c++0x. Based on what you mentioned, changing it to c++03 should work as well. I think this might be a useful reference for someone who is in the similar situation as me (an old version of g++, new version of matlab, permission limits and not very familiar with mex).
